### ISSUE TITLE
onHttpCallResponse in Root Context getting blocked till after onStart of other VMs

### ISSUE THREAD

**Issue by mcieplak** [envoyproxy/envoy-wasm#599]
### Problem Statement
We are building a Wasm based C++ Envoy filter using proxy-wasm-cpp-sdk that needs to make an external HTTP call before the stream context starts. However we are running into some issues (described below) when there is a delayed response from the server. We are running the setup in Istio 1.5.5.

### Current Implementation
In order to ensure that the data gets populated prior to servicing requests our plan was to setup a singleton Wasm VM which does the following in its [onStart()](https://gist.github.com/mcieplak/b0d11d420bac4a32ca930b16e888b976) function:
1. Makes an httpCall to the external server 
2. Sets the shared data with http response json value

We then have the non singleton VMs which execute the following logic in their [onStart()](https://gist.github.com/mcieplak/b3e0352e534fa763fa7be712011b7aef) function:
1. Wait for the shared data to be present for upto 10 secs 
2. Read the shared data value and load it into memory

### Test case
1. Set the Envoy worker thread concurrency value to 2
2. Add a delay of 5 seconds to the response from the HTTP external server

### Expected Behavior
- 1 BaseVM executing the singleton wasm module
- 2 Thread-Local VMs executing the filter wasm module
- In RootContext onStart() function Singleton VM fetches the data and writes to the shared buffer.
- In RootContext onStart() Thread-local VMs read the data from the buffer and keep it in memory.
- Thread-Local VMs continue the stream context based execution.

### Observed behavior
- 1 Base VM executing the singleton wasm module appearing to be part of the main Envoy thread
- 1 Base VM and 1 Thread-Local VM executing the filter wasm module both appearing to be part of the same main Envoy thread
- 2 Thread-Local VMs executing the filter wasm module appearing to be part of the Envoy worker threads
- The Base VM and Thread-Local VM part of the main worker thread are unable to read the shared data and fail in their onStart()
- The onHttpCallResponse fn gets called after the failure of the VMs in the main worker thread
- The 2 thread local VMs are able to then read from the shared data
- The Base VMs are not invoked when request are made to the envoy proxy

Based on the logs (as shown below) it seems as if the onStart functions of the VMs booted as part of the main Envoy thread need to complete before the onHttpCallResponse can be processed. This occurs only in the case of the external HTTP server taking a longer time to respond.

### Questions
1. Is this the correct approach in order to achieve our objective?
2. What is the difference between Base VM and Thread-local VM?
3. Why are there Base VMs created in addition to the Thread-local VMs for the filter wasm module that is meant to be run only per worker thread?
4. From within a VM is it possible to identify if the given VM is a Base or a thread local VM?

### Log Output
<pre>
[Envoy (Epoch 0)] [2020-08-03 21:10:21.114][19][debug][wasm] [bazel-out/k8-opt/bin/external/envoy/source/extensions/common/wasm/_virtual_includes/wasm_vm_base/extensions/common/wasm/wasm_vm_base.h:38] WasmVm created envoy.wasm.runtime.v8 now active
[Envoy (Epoch 0)] [2020-08-03 21:10:21.114][19][debug][wasm] [external/envoy/source/extensions/common/wasm/wasm.cc:121] Base Wasm created 1 now active
<b>[Envoy (Epoch 0)] [2020-08-03 21:10:21.274][19][warning][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1092] wasm log my_filter my_filter: secret fetch httpCall was successful</b>
... 
[Envoy (Epoch 0)] [2020-08-03 21:10:22.842][19][debug][config] [external/envoy/source/extensions/filters/network/http_connection_manager/config.cc:420]       name: envoy.filters.http.wasm
   "runtime": "envoy.wasm.runtime.v8",
     "filename": "/var/local/lib/wasm-filters/my_filter.wasm"
<b>[Envoy (Epoch 0)] [2020-08-03 21:10:22.845][19][debug][wasm] [bazel-out/k8-opt/bin/external/envoy/source/extensions/common/wasm/_virtual_includes/wasm_vm_base/extensions/common/wasm/wasm_vm_base.h:38] WasmVm created envoy.wasm.runtime.v8 now active
[Envoy (Epoch 0)] [2020-08-03 21:10:22.845][19][debug][wasm] [external/envoy/source/extensions/common/wasm/wasm.cc:121] Base Wasm created 6 now active
[Envoy (Epoch 0)] [2020-08-03 21:10:22.932][19][debug][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1086] wasm log my_filter my_filter: Sleeping for 1 second...</b>
[Envoy (Epoch 0)] [2020-08-03 21:10:23.981][19][debug][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1086] wasm log my_filter my_filter: Sleeping for 1 second...
...
[external/envoy/source/extensions/common/wasm/context.cc:1086] wasm log my_filter my_filter: Sleeping for 1 second...
<b>[Envoy (Epoch 0)] [2020-08-03 21:10:33.423][19][debug][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1086] wasm log my_filter my_filter: Getting shared data secret failed
[Envoy (Epoch 0)] [2020-08-03 21:10:33.423][19][debug][wasm] [bazel-out/k8-opt/bin/external/envoy/source/extensions/common/wasm/_virtual_includes/wasm_vm_base/extensions/common/wasm/wasm_vm_base.h:38] WasmVm created envoy.wasm.runtime.v8 now active
[Envoy (Epoch 0)] [2020-08-03 21:10:33.444][19][debug][wasm] [external/envoy/source/extensions/common/wasm/wasm.cc:285] Thread-Local Wasm created 7 now active</b>
[Envoy (Epoch 0)] [2020-08-03 21:10:33.446][19][debug][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1086] wasm log my_filter my_filter: Sleeping for 1 second...
...
[Envoy (Epoch 0)] [2020-08-03 21:10:42.884][19][debug][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1086] wasm log my_filter my_filter: Sleeping for 1 second...
<b>[Envoy (Epoch 0)] [2020-08-03 21:10:43.932][19][debug][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1086] wasm log my_filter my_filter: Getting shared data secret failed
[Envoy (Epoch 0)] [2020-08-03 21:10:43.960][19][warning][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1092] wasm log my_filter: Secret fetch successful
[Envoy (Epoch 0)] [2020-08-03 21:10:43.982][30][debug][wasm] [bazel-out/k8-opt/bin/external/envoy/source/extensions/common/wasm/_virtual_includes/wasm_vm_base/extensions/common/wasm/wasm_vm_base.h:38] WasmVm created envoy.wasm.runtime.v8 now active
[Envoy (Epoch 0)] [2020-08-03 21:10:43.991][30][debug][wasm] [external/envoy/source/extensions/common/wasm/wasm.cc:285] Thread-Local Wasm created 10 now active
[Envoy (Epoch 0)] [2020-08-03 21:10:43.995][30][debug][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1086] wasm log my_filter my_filter: Secret json parsed successfully
[Envoy (Epoch 0)] [2020-08-03 21:10:43.996][30][debug][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1086] wasm log my_filter my_filter: Converting shared data to secret json was successful</b>
...
<b>[Envoy (Epoch 0)] [2020-08-03 21:10:44.011][31][debug][wasm] [bazel-out/k8-opt/bin/external/envoy/source/extensions/common/wasm/_virtual_includes/wasm_vm_base/extensions/common/wasm/wasm_vm_base.h:38] WasmVm created envoy.wasm.runtime.v8 now active
[Envoy (Epoch 0)] [2020-08-03 21:10:44.100][31][debug][wasm] [external/envoy/source/extensions/common/wasm/wasm.cc:285] Thread-Local Wasm created 13 now active
[Envoy (Epoch 0)] [2020-08-03 21:10:44.105][31][debug][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1086] wasm log my_filter my_filter: Secret json parsed successfully
[Envoy (Epoch 0)] [2020-08-03 21:10:44.105][31][debug][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1086] wasm log my_filter my_filter: Converting shared data to secret json was successful</b>
</pre>

cc @jplevyak @PiotrSikora

**Comment by jplevyak** [comment]
Would it be possible to attach a (perhaps cut down) version of your configuration and/or plugin?

It would help ensure that any changes fix what you are seeing.

**Comment by mcieplak** [comment]
@jplevyak Sure, here is the code we're using:
Singleton filter: https://gist.github.com/mcieplak/b0d11d420bac4a32ca930b16e888b976
Non singleton filter: https://gist.github.com/mcieplak/b3e0352e534fa763fa7be712011b7aef

**Comment by mandarjog** [comment]
@jplevyak @kyessenov 
We should not busy wait in the main thread, othewise we are not giving the other thread to finish its work.
I am wondering if the non-singleton filter should try to read the the shared data in the stream context, not the main context.
If it cannot read the data, the filter should treat it like NOOP, or propagation delay. It will only have to read it once and store it in the root context.

**Comment by jplevyak** [comment]
You cannot hard loop in Envoy under any circumstances.  Envoy is an event driven system and each worker thread is shared with many many connections.  This is not a Wasm problem.  Please revise your code and if you have problems open a new issue.
